src/lowlevel/connection.rs

```diff
@@ -63,36 +70,27 @@ impl<W, Buffer, Auxiliary> Drop for SharedData<W, Buffer, Auxiliary> {
 }
 
 impl<W, Buffer, Auxiliary> SharedData<W, Buffer, Auxiliary> {
-    pub(crate) fn writer(&self) -> &WriterBuffered<W> {
-        &self.0.writer
+    pub(crate) fn writer(&self) -> Pin<&WriterBuffered<W>> {
+        self.0.as_ref().project_ref().writer
     }
 
     pub(crate) fn responses(&self) -> &AwaitableResponses<Buffer> {
         &self.0.responses
     }
 
-    pub(crate) fn get_mut_writer(&mut self) -> Option<&mut WriterBuffered<W>> {
-        Arc::get_mut(&mut self.0).map(|shared_data| &mut shared_data.writer)
-    }
-
     /// `SharedData` is a newtype wrapper for `Arc<SharedDataInner>`,
     /// so this function returns how many `Arc` there are that referred
     /// to the shared data.
     #[inline(always)]
     pub fn strong_count(&self) -> usize {
-        Arc::strong_count(&self.0)
+        pinned_arc_strong_count(&self.0)
```

Comment:
I don't understand why we even provide the `strong_count` function? It's basically _never_ what you want to use, since most attempts to make decisions based on the strong count are subject to race conditions.

Reply:
It's used in [`Drop::drop`](https://github.com/openssh-rust/openssh-sftp-client/blob/fab8822ad3c9b2875013480a165c9e98f546b3e7/src/lowlevel/connection.rs#L48) implementation for `SharedData`.

When there is only last `SharedData` left, it will set `is_conn_closed` to `true` and `notify_read_end` so that [`SharedData::wait_for_new_request`](https://github.com/openssh-rust/openssh-sftp-client/blob/fab8822ad3c9b2875013480a165c9e98f546b3e7/src/lowlevel/connection.rs#L113).

Reply:
One way to fix this is to add another atomic field to keep track of the number of `WriteEnd`.

Reply:
Another alternative would be to use [`tokio::sync::mpsc`](https://docs.rs/tokio/latest/tokio/sync/mpsc) with `T = ()`, but I suspect it will be much more heavy than `Notify`.

Checking its implementation, it uses [mod `chan`](https://docs.rs/tokio/latest/src/tokio/sync/mpsc/chan.rs.html), which contains `Arc<Chan<T>>` that uses `Notify` and `Semaphore` internally.

Reply:
> I don't understand why we even provide the strong_count function? It's basically never what you want to use, since most attempts to make decisions based on the strong count are subject to race conditions.

I agree this should be a private function though, would fix it.

Reply:
Done!

It is now `pub(crate)` only.

Reply:
Hmm, that implementation looks racy to me if we ever give out the underlying `Arc`. All it takes is for someone to downgrade that `Arc` to a `Weak`, then call `drop` while racing that against a `Weak::upgrade`. If we never give out the actual `Arc` (so no downgrade is possible), *and* we _never_ clone the `Arc` from a `ReadEnd`, then I suppose it's probably fine.

Reply:
> Hmm, that implementation looks racy to me if we ever give out the underlying `Arc`. All it takes is for someone to downgrade that `Arc` to a `Weak`, then call `drop` while racing that against a `Weak::upgrade`. If we never give out the actual `Arc` (so no downgrade is possible), _and_ we _never_ clone the `Arc` from a `ReadEnd`, then I suppose it's probably fine.

The users can never access to the underlying Arc, it can only deal with SharedData, WriteEnd and ReadEnd, and ReadEnd never cloned the SharedData/underlying Arc or using the weak pointer.

Reply:
And, I assume, a `ReadEnd` can't itself be cloned?

I would probably put that entire argument for correctness in the place where we currently check the strong count.

Reply:
No, there can be only one ReadEnd.

Reply:
> I would probably put that entire argument for correctness in the place where we currently check the strong count.

Sure!

Reply:
Done!